lib/tokenizer/rainscriptTokenizer.js

```diff
@@ -0,0 +1,72 @@
+import Lexer from 'lex';
+import { _ } from 'underscore';
+import { rainscriptTokenizerRules } from './rainscriptTokenizerRules';
+
+const ccount = require('ccount');
+
+// the regex-based lexer for the Rainscript tokenizer
+let _lexer = new Lexer(token(null));
+
+// grab rules from tokenizerRules file
+rainscriptTokenizerRules.forEach(rule =>
+  _lexer.addRule(rule.pattern, token(rule.type), rule.start)
+);
+
+/**
+ * A factory method that creates a new Rainscript tokenizer instance.
+ * @returns an instance of Rainscript's tokenizer
+ */
+export const rainscriptTokenizer = () => {
+  // wrap lexer prototype so that tokenizers can be thread-safe
+  let lexer = Object.create(_lexer);
```

Comment:
why `Object.create(...)` ???

Reply:
## Quick Explanation
Main reason for using `Object.create(...)` is to avoid potential data races for this line:

```js
34      lexer.setInput(text);
```

## Long Explanation

Not wrapping `_lexer` with `Object.create(...)` will pose some problems when there are multiple instances of `rainscriptTokenizer`. This is primarily because `_lexer` is a shared instance. Hence, it must be wrapped in a prototype so that instance-specific properties don't leak.

### Prototype Chain

As an illustration, here is the prototype chain of `_lexer`:
```js 
_lexer = this {
  state: 0,
  index: 0,
  input: '',
  addRule: [Function],
  setInput: [Function],
  lex: [Function],
}
-> Lexer {}
-> Object {}
-> null
```

Now, observe how `addRule` and `setInput` are direct properties of `this`. Usually, they should be located in the `Lexer` prototype, but for some reason, this is not the case for the `lex` package (most likely because it was written pre-EcmaScript).

So what if it's directly in `this`? Well, the consequence is that whenever one calls `_lexer.setInput(...)`, the variable `this.input`, not to mention `this.index`, will be manipulated directly. Meaning, `_lexer` becomes a _shared instance_. Data won't be tokenized correctly as expected If passed around as is.

### Data Races

Suppose we don't wrap `_lexer` as a prototype. Consider running two tests asynchronously:

```js
function firstTest() {
  let tokenizer = rainscriptTokenizer();
  let tokens = tokenizer.tokenize('Hello');
  return tokens;
}

function secondTest() {
  let tokenizer = rainscriptTokenizer();
  let tokens = tokenizer.tokenize('Goodbye');
  return tokens;
}

setTimeout(firstTest, 10);
setTimeout(secondTest, 10);
```

In the case above, there's no guarantee that `firstTest` finishes before `secondTest`. But note that by assumption, both functions have tokenizers that point to the same `_lexer` instance. Now, we have this potential conflict:

firstTest | secondTest | `_lexer.input`
-------------|-------------|---------------
`_lexer.setInput('Hello')` | ... | 'Hello'
... | `_lexer.setInput('Goodbye')`  | 'Goodbye'
... | `let token = _lexer.lex();`  | `Goodbye'
`let token = _lexer.lex();` | ...  | **ERROR**

If the order of execution is above, the last iteration will throw an **ERROR** since by that time, `_lexer.input` already contained `Goodbye`. Whelp.

### Object.create(_lexer)

How to overcome this? We can wrap the `_lexer` instance with `Object.create(...)`:

```js
let lexer = Object.create(_lexer);
```

According to ES6 [reference](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/create):

> The `Object.create()` method creates a new object with the specified prototype object and properties.

In short, the above code will create a new `lexer` instance with prototype `_lexer`. In effect, we get the following prototype chain:

```js
lexer = this {
  state: 0,
  index: 0,
  input: ''
}
-> _lexer {
  addRule: [Function],
  setInput: [Function],
  lex: [Function],
}
-> Lexer {}
-> Object {}
-> null
```

This is clearer to work with. Calling `lexer.setInput(...)` now sets the `this.input` correctly without manipulating the shared `_lexer`. There should be no data races anymore here.

Furthermore, we get to recycle `addRule`, `setInput`, and `lex` without having to create a new `_lexer` all over again.

### Why not clone instead?

Actually, we can also get the same functionality by cloning the properties of `_lexer` to another object. But unlike cloning, `Object.create` allows changes to be made to the `_lexer` prototype without having to make changes to all `lexer` instances. Either way, `Object.create` is more compact and recommended by ES6 factory pattern, so it is preferred over cloning nonetheless.